q2_metadata/normalization/_norm_check_rules.py

```diff
@@ -0,0 +1,311 @@
+# ----------------------------------------------------------------------------
+# Copyright (c) 2016-2020, QIIME 2 development team.
+#
+# Distributed under the terms of the Modified BSD License.
+#
+# The full license is in the file LICENSE, distributed with this software.
+# ----------------------------------------------------------------------------
+
+from q2_metadata.normalization._norm_errors import (
+    RuleError, ExpectedError, OntologyError, RemapError,
+    ValidationError, NormalizationError,
+    BlankError, MissingError, FormatError
+)
+
+
+def check_rule(variable: str, rules: dict, rule: str, rule_value):
+    """For each rule in the current variable's rules set,
+    perform checks that the rule values to be used for the rules
+    to apply are properly formatted.
+        - If properly formatted: full the rules dict
+        for the variable's Rules() class instance.
+        - If not properly formatted: raise a specific error and
+        collect the encountered error in the ErrorsCollection()
+        instance for this variable's Rules() class instance.
+
+    Parameters
+    ----------
+    variable : str
+        Name of the metadata variable for which there is rules.
+    rules : dict
+        Rules in the hard-coded data structure that has
+        common default values for all variables and
+        that is checked and updated here.
+    rule : str
+        Current rule name. Could be one of these:
+        - expected: accepted, controlled vocabulary from AGP.
+        - ontology: accepted, controlled vocabulary from a third-party resource.
+        - remap: the replacements to perform ( "A" -> "B" )
+        - validation: the
+        - normalization:
+        - blank:
+        - missing:
+        - format:
+    rule_value : str, list or dict
+        Parsed rule value for the current rule.
+    """
+    rule_type_map = {
+        'expected': 'lookups',
+        'ontology': 'lookups',
+        'remap': 'edits',
+        'validation': 'edits',
+        'normalization': 'edits',
+        'blank': 'allowed',
+        'missing': 'allowed',
+        'format': 'allowed'
+    }
+
+    if rule not in rule_type_map:
+        raise RuleError(rule, variable, rule_value)
+
+    if rule in ['blank', 'missing']:
+        rule_method_key = 'check_blank_missing'
+    else:
+        rule_method_key = 'check_%s' % rule
+
+    rule_method = globals()[rule_method_key]
+    rule_error = globals()['%sError' % rule.capitalize()]
```

Comment:
It is in general not a good idea to use global variable references. Is there a reason why you can't use the class itself as an argument?

Reply:
no reason - great idea: will do.
I always wondered how to call a function using it's character string "name". You seem to indicate that this is possible through class method call. In this case I will have to roll back to a longer dictionary of class.method (value) per rule to check (key).

Reply:
Yes you are both right - and that was the solution originally started by Rebecca.

Reply:
because in the latest version the check also collect the errors (without raising error : user-defined error classes have been deleted), this part now is:

```
        rule_type_map = {
            'normalization': (self.check_normalization, 'edits'),
            'validation': (self.check_validation, 'edits'),
            'remap': (self.check_remap, 'edits'),
            'expected': (self.check_expected, 'lookups'),
            'ontology': (self.check_allowed, 'lookups'),
            'blank': (self.check_allowed, 'allowed'),
            'missing': (self.check_allowed, 'allowed'),
            'format': (self.check_allowed, 'allowed')
        }
        if rule in rule_type_map:
            rule_method, rule_type = rule_type_map[rule]
            error = rule_method(rule_value, self.rules_allowed[rule])
            if not len(error):
                self.variables_rules[
                    variable].rules[rule_type][rule] = rule_value
            else:
                self.rules_format_error.errors.append(
                    [variable, rule, rule_value,
                     rule_type, error[0], error[1]]
                )
        else:
            self.rules_format_error.errors.append(
                [variable, rule, rule_value,
                 '', 'rule not recognized', '']
            )

```